### PR TITLE
Offer an integer math-based emission curve

### PR DESCRIPTION
**Pull request by CjS77**
PResented is an emission curve algorithm that
* uses only integer math
* can approximate the current floating point algorithm to an arbitrary
degree of accuracy

Closed form solutions of the reward and supply are not possible with
this approach, so the `block_reward` and `total_supply` functions are
removed. Only the iterator is used to determine emission.

### CODE REVIEW

**Comment by CjS77** [comment]
Some things to note about this implementation:

A way of thinking about this is `k * A` where k <~ 1  is rewritten (roughly) as ` A - trunc((1-k) * A)`.

Once A is less than the lowest precision of `(1-k)`, the right hand term becomes zero. 

e.g. if k = 0.99999, (1-k) = 1e-6 the lowest precision is -6, so if A < 1e6, the sum is 
```
A - trunc(1e-6 * 0.99999e6) = A - trunc(0.999) = A
```

This is basically an implicit tail emission, so a simple adjustment to give a more intuitive result is to change

```rust
    fn next_reward(&self) -> MicroTari {
        let r: u64 = self.reward.into();
        self.schedule
            .decay
            .iter()
            .fold(self.schedule.tail + self.reward, |sum, i| {
                sum - MicroTari::from(r >> *i)
            })
    }
```

to

```rust
fn next_reward(&self) -> MicroTari {
        let r: u64 = self.reward.into();
        let next = self.schedule
            .decay
            .iter()
            .fold(self.reward, |sum, i| {
                sum - MicroTari::from(r >> *i)
            });
      if next > self.schedule.tail {
          next
     } else {
         self.schedule.tail
    }
}
```

**Review by SWvheerden — DISMISSED** [review]
Looks good. 

I agree with the proposed changed forward, it will make it more clear whats going on.

**Comment by CjS77** [comment]
Ugh, I read somewhere recently that `1 >> n` where `n>1` on x86 chips causes an overflow, which may have implications for this approach

File: base_layer/core/src/consensus/emission.rs [hunk 1]
```diff
@@ -107,14 +107,24 @@ impl<'a> EmissionRate<'a> {
         self.reward
     }
 
+    /// Calculates the next reward by multiplying the decay factor by the previous block reward using integer math.
+    ///
+    /// We write the decay factor, 1 - k, as a sum of fraction powers of two. e.g. if we wanted 0.25 as our k, then
+    /// (1-k) would be 0.75 = 1/2 plus 1/4 (1/2^2).
+    ///
+    /// Then we calculate k.R = (1 - e).R = R - e.R = R - (0.5 * R + 0.25 * R) = R - R >> 1 - R >> 2
```
Comment (stringhandler) [inline base_layer/core/src/consensus/emission.rs:115]:
```suggestion
    /// Then we calculate k.R = (1 - e).R = R - e.R = R - (0.5 * R + 0.25 * R) = R - R >> 1 - R >> 2
    /// If the value returned is never lower than the tail emission
```

**Review by stringhandler — APPROVED** [review]
Agree in concept and approach. This is a little different from the previous emission where the tail emission was added, instead of `max(reward, tail)`, but that is fine and the final emission for mainnet has not yet been decided. 
I haven't tested it locally, but can approve from code review. I added a suggestion for an extra doc comment

File: base_layer/core/src/consensus/emission.rs (hunk 1 above)
Comment (stringhandler) [reply base_layer/core/src/consensus/emission.rs:115]: Lol. Should be `The value return is never lower than the tail emission`

**Pull request merged**